### PR TITLE
indicator to SOS bridge

### PR DESCRIPTION
**Pull request by matbesancon**
Based on the comment on indicator constraints in the [SCIP documentation](https://scip.zib.de/doc/html/cons__indicator_8h.php):
> This constraint is equivalent to a linear constraint `ax−s ≤ b` and an SOS1 constraint on z and s (at most one should be nonzero). In the indicator context we can, however, separate more inequalities.

### CODE REVIEW

**Comment by matbesancon** [comment]
this should be good for review

**Comment by matbesancon** [comment]
There seems to be a problem, but I haven't figured out what exactly. I still haven't figured how to check exactly the newly formulated problem

**Comment by odow** [comment]
> I still haven't figured how to check exactly the newly formulated problem

Try saving to MathOptFormat?

**Comment by matbesancon** [comment]
Ah yes thanks, I was looking for an excuse to try it :)

File: src/Bridges/Constraint/indicator.jl [hunk 1]
```diff
@@ -51,3 +51,89 @@ function concrete_bridge_type(::Type{<:IndicatorActiveOnFalseBridge},
                               ::Type{MOI.IndicatorSet{MOI.ACTIVATE_ON_ZERO, S}}) where {F<:MOI.VectorAffineFunction, S<:MOI.AbstractScalarSet}
     return IndicatorActiveOnFalseBridge{Float64, F, S}
 end
+
+## Indicator constraint by SOS1 bridge
+
+"""
```
Comment (blegat) [inline src/Bridges/Constraint/indicator.jl]: I would move it to another file `indicator_sos1.jl`

File: src/Bridges/Constraint/Constraint.jl [hunk 2]
```diff
@@ -64,12 +64,12 @@ const RSOCtoPSD{T, OT<:MOI.ModelLike} = SingleBridgeOptimizer{RSOCtoPSDBridge{T}
...
 
 """
-    add_all_bridges(bridged_model, T::Type)
+    add_all_bridges(bridged_model, ::Type{T})
 
 Add all bridges defined in the `Bridges.Constraint` submodule to
 `bridged_model`. The coefficient type used is `T`.
 """
-function add_all_bridges(bridged_model, T::Type)
+function add_all_bridges(bridged_model, ::Type{T}) where {T}
```
Comment (blegat) [inline src/Bridges/Constraint/Constraint.jl]: Why this change ? Is it better like this ?

File: src/Bridges/Constraint/indicator.jl [hunk 3]
```diff
@@ -51,3 +51,89 @@ function concrete_bridge_type(::Type{<:IndicatorActiveOnFalseBridge},
...
+The `IndicatorSOS1Bridge` replaces an indicator constraint of the following form:
+``z \\in \\mathbb{B}, z == 1 \\implies f(x) \\leq b`` with a SOS1 constraint:
+``z \\in \\mathbb{B}, w \\leq 0, f(x) + w \\leq b, SOS1(w, z)``.
+`GreaterThan` constraints are handled in a symmetric way:
+``z \\in \\mathbb{B}, z == 1 \\implies f(x) \\geq b`` is reformulated as:
+``z \\in \\mathbb{B}, w \\geq 0, f(x) + w \\geq b, SOS1(w, z)``.
+`EqualTo` constraints are handled without a bound constraint:
+``z \\in \\mathbb{B}, z == 1 \\implies f(x) == b`` is reformulated as:
+``z \\in \\mathbb{B}, w \\text{ free}, f(x) + w == b, SOS1(w, z)``.
+
```
Comment (blegat) [inline src/Bridges/Constraint/indicator.jl]: Why not always add a free `w` so that it works for any indicator constraint (also `MOI.Interval`, ...) ?

File: src/Bridges/Constraint/Constraint.jl (hunk 2 above)
Comment (matbesancon) [reply src/Bridges/Constraint/Constraint.jl]: `T::Type`, using the run-time value, which is a type.
`::Type{T}`: dispatching on `T`, which is a compile-time info

File: src/Bridges/Constraint/indicator.jl (hunk 3 above)
Comment (matbesancon) [reply src/Bridges/Constraint/indicator.jl]: Can we have a `free` fallback with all constraints but `LessThan`, `GreaterThan`, otherwise we add a free, which has to be re-bounded by two variables for `LP`& `MILP` solvers

File: src/Bridges/Constraint/Constraint.jl (hunk 2 above)
Comment (blegat) [reply src/Bridges/Constraint/Constraint.jl]: Do you have a reference for that ? I thought I read that it recognizes this special case and it's equivalent

File: src/Bridges/Constraint/indicator.jl (hunk 3 above)
Comment (blegat) [reply src/Bridges/Constraint/indicator.jl]: Not sure I understood "which has to be re-bounded by two variables"

Comment (matbesancon) [reply src/Bridges/Constraint/indicator.jl]: Sorry yes, if we create the variable `w` free, it will be replaced by `w+ >= 0, w- >= 0, w = w+ + w- ` instead of creating a bounded `w` right away

File: src/Bridges/Constraint/Constraint.jl (hunk 2 above)
Comment (matbesancon) [reply src/Bridges/Constraint/Constraint.jl]: so it's a special parsing rule for the compiler? I wasn't aware of one

File: src/Bridges/Constraint/indicator.jl (hunk 3 above)
Comment (blegat) [reply src/Bridges/Constraint/indicator.jl]: That's also the case for SDP solvers with the `Variable.FreeBridge`. Which LP solver does not support free variables ? This transformation is done internally ?
For a solver supporting free variables such constraints may increase the size of the problem, we should probably have *informative* constraints but that's out of the scope of this PR.

Comment (matbesancon) [reply src/Bridges/Constraint/indicator.jl]: > Which LP solver does not support free variables ? This transformation is done internally ?

All solvers (that I am aware of) support free variables, but the transformation needs to be done internally somehow. The simplex algorithm does require positive variables

File: src/Bridges/Constraint/Constraint.jl (hunk 2 above)
Comment (blegat) [reply src/Bridges/Constraint/Constraint.jl]: Not sure but I think the compiler might treat `Type` variables differently than other variables. Having to write `::Type{T}` everytime an argument is a type instead of just `T` or `T::Type` seems like working around something a compiler might do itself

File: src/Bridges/Constraint/indicator.jl (hunk 3 above)
Comment (blegat) [reply src/Bridges/Constraint/indicator.jl]: If you want to use the standard form you have equality constraints and nonnegative variables but then the dual has free variables. Some solvers might solve the dual using the simplex in which case they need all variables to be free. Although that's the textbook simplex so I don't know whether solvers might use a more sophisticated one supporting mixing free and non-free ones.

Comment (matbesancon) [reply src/Bridges/Constraint/indicator.jl]: > If you want to use the standard form you have equality constraints and nonnegative variables but then the dual has free variables.

My understanding is that the standard form is a form  `A x <= b, x >= 0` transformed in `Ax + s = b, x, s >= 0`. You need to have basic columns in the equality, so starting from a generic `A x = b`, it will need to change the variable system in `[A; -A] x <= [b; -b]` and then add slacks, in order to have as many basic columns as constraints.

File: test/Bridges/Constraint/indicator.jl [hunk 4]
```diff
@@ -53,3 +53,92 @@ include("../utilities.jl")
         @test t.coefficient ≈ 1.0
     end
 end
+
+@testset "Indicator by SOS1" begin
+    # linear problem with indicator constraints
```
Comment (blegat) [inline test/Bridges/Constraint/indicator.jl]: Can you also split the file in two so that we have the same files in `src` and `test` folders ?

File: test/Bridges/Constraint/indicator.jl [hunk 5]
```diff
@@ -53,3 +53,92 @@ include("../utilities.jl")
         @test t.coefficient ≈ 1.0
     end
 end
+
```
Comment (blegat) [inline test/Bridges/Constraint/indicator.jl]: We should also have a test with `basic_constraint_test` (see https://github.com/JuliaOpt/MathOptInterface.jl/issues/750) and a test against `indicator1`, `indicator2` or `indicator3` (https://github.com/JuliaOpt/MathOptInterface.jl/blob/98a0e472e1757db14b849bbba8fe84597856def2/src/Test/intlinear.jl#L609-L611) with `test_model_equals` (see https://github.com/JuliaOpt/MathOptInterface.jl/issues/820)

File: src/Bridges/Constraint/indicator.jl (hunk 1 above)
Comment (matbesancon) [reply src/Bridges/Constraint/indicator.jl]: done in last commit

File: test/Bridges/Constraint/indicator.jl (hunk 4 above)
Comment (matbesancon) [reply test/Bridges/Constraint/indicator.jl]: Done in last commit

Comment (matbesancon) [reply test/Bridges/Constraint/indicator.jl]: of course not, it's much more fun when it all stays local :upside_down_face:

Comment (matbesancon) [reply test/Bridges/Constraint/indicator.jl]: (just did)

File: test/Bridges/Constraint/indicator.jl (hunk 5 above)
Comment (matbesancon) [reply test/Bridges/Constraint/indicator.jl]: I didn't manage to make that work (yet)

**Comment by matbesancon** [comment]
@blegat I can't figure out what is off here. Any tips on a debugging process for that?

**Comment by blegat** [comment]
> I can't figure out what is off here. Any tips on a debugging process for that?

You mean for making `basic_constraint_tests` working ? What is the issue you are facing ? Can you share an error message ?

**Comment by matbesancon** [comment]
The constraints defined in the bridge seem not to be activated.

**Comment by matbesancon** [comment]
> You mean for making basic_constraint_tests working ?

Or even when testing the bridge with `Cbc`

**Comment by blegat** [comment]
So the indicator constraint is added but the underlying model does not receive the SOS1 constraints ?
I would be good to add a test for passing the indicator test with a `Bridges.Constraint.IndicatortoSOS1`

**Comment by matbesancon** [comment]
> So the indicator constraint is added but the underlying model does not receive the SOS1 constraints ?

Yes I assume that's it

>  I would be good to add a test for passing the indicator test with a Bridges.Constraint.IndicatortoSOS1

Agreed, that's what I was trying to figure out

File: src/Bridges/Constraint/indicator_sos.jl [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+function bridge_constraint(::Type{IndicatorSOS1Bridge{T,BC,MaybeBC}}, model::MOI.ModelLike, f::MOI.VectorAffineFunction{T}, s::MOI.IndicatorSet{MOI.ACTIVATE_ON_ONE, BC}) where {T <: Real, BC, MaybeBC}
+    f_scalars = MOIU.eachscalar(f)
+    (w, bound_constraint) = _add_bound_constraint!(model, BC)
+    z = f_scalars[1].terms[1].variable_index
+    sos_vector = MOI.VectorOfVariables([w, z])
+    sos_constraint = MOI.add_constraint(model, sos_vector, MOI.SOS1{T}([0.4, 0.6]))
+    affine_func = copy(f_scalars[2])
+    push!(
+        affine_func.terms,
+        MOI.ScalarAffineTerm{T}(one(T), w)
```
[the remark this answers is not in the document]
Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: done. I feel like this function is `MOIU.operate!` is not the best design, it's really obscure to read

File: src/Bridges/Constraint/indicator_sos.jl [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    return IndicatorSOS1Bridge{T, S, MOI.ConstraintIndex{MOI.SingleVariable, S}}
+end
+
+function concrete_bridge_type(::Type{<:IndicatorSOS1Bridge{T}},
+                              ::Type{<:MOI.VectorAffineFunction},
+                              ::Type{MOI.IndicatorSet{MOI.ACTIVATE_ON_ONE, S}}) where {T, S <: MOI.AbstractScalarSet}
+    return IndicatorSOS1Bridge{T, S, Nothing}
+end
+
+function concrete_bridge_type(::Type{<:IndicatorSOS1Bridge},
```
[the remark this answers is not in the document]
Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: done

File: src/Bridges/Constraint/Constraint.jl [hunk 8]
```diff
@@ -61,15 +61,16 @@ const RootDet{T, OT<:MOI.ModelLike} = SingleBridgeOptimizer{RootDetBridge{T}, OT
 include("soc_to_psd.jl")
 const SOCtoPSD{T, OT<:MOI.ModelLike} = SingleBridgeOptimizer{SOCtoPSDBridge{T}, OT}
 const RSOCtoPSD{T, OT<:MOI.ModelLike} = SingleBridgeOptimizer{RSOCtoPSDBridge{T}, OT}
-include("indicator.jl")
+include("indicator_activate_on_zero.jl")
+include("indicator_sos.jl")
```
[the remark this answers is not in the document]
Comment (matbesancon) [reply src/Bridges/Constraint/Constraint.jl]: done

File: src/Bridges/Constraint/indicator_sos.jl [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+end
+
+function bridge_constraint(::Type{IndicatorSOS1Bridge{T,BC,MaybeBC}}, model::MOI.ModelLike, f::MOI.VectorAffineFunction{T}, s::MOI.IndicatorSet{MOI.ACTIVATE_ON_ONE, BC}) where {T <: Real, BC, MaybeBC}
+    f_scalars = MOIU.eachscalar(f)
+    (w, bound_constraint) = _add_bound_constraint!(model, BC)
+    z = convert(MOI.SingleVariable, f_scalars[1]).variable
+    sos_vector = MOI.VectorOfVariables([w, z])
+    sos_constraint = MOI.add_constraint(model, sos_vector, MOI.SOS1{T}([0.4, 0.6]))
+    affine_func = f_scalars[2]
+    MOIU.operate!(+, T, affine_func, MOI.SingleVariable(w))
```
Comment (blegat) [inline src/Bridges/Constraint/indicator_sos.jl]: If `f` is `VectorOfVariables`, `affine_func` is `SingleVariable` and it cannot hold the sum so to be safe, do `some_name = MOIU.operate!(...`

File: src/Bridges/Constraint/indicator_sos.jl [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+function _add_bound_constraint!(model::MOI.ModelLike, ::Type{BC}) where {T <: Real, BC <: Union{MOI.LessThan{T}, MOI.GreaterThan{T}}}
+    return MOI.add_constrained_variable(model, BC(zero(T)))
+end
+
+function _add_bound_constraint!(model::MOI.ModelLike, ::Type{<:MOI.AbstractScalarSet})
+    return (MOI.add_variable(model), nothing)
+end
+
+function MOI.supports_constraint(::Type{<:IndicatorSOS1Bridge},
+                                 ::Type{<:MOI.VectorAffineFunction},
```
[the remark this answers is not in the document]
Comment (blegat) [reply src/Bridges/Constraint/indicator_sos.jl]: You don't have to check that the solver supports the added constraints. This is done automatically by the bridge optimizer using the `added_...` functions.

File: src/Bridges/Constraint/indicator_sos.jl (hunk 9 above)
Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: done

**Comment by matbesancon** [comment]
Still no idea about the segfault on 1.0

**Comment by blegat** [comment]
> Still no idea about the segfault on 1.0

You are probably just unlucky, I would recommend finishing the PR first and then see if the segfault is still there. It might disappear randomly just like it appeared.
For instance, addressing https://github.com/JuliaOpt/MathOptInterface.jl/pull/877#discussion_r341394561 will affect methods of `supports_constraint` so it might affect the segfault situation.

File: src/Bridges/Constraint/indicator_sos.jl (hunk 10 above)
[the remark this answers is not in the document]
Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: done

**Comment by matbesancon** [comment]
re-opening, there was a HTTP connection error

**Comment by matbesancon** [comment]
ok, back to square one, only Julia 1.0 is failing

**Comment by matbesancon** [comment]
> src/Test/UnitTests/basic_constraint_tests.jl

Finally added it, I defined the corresponding `MOI.get` methods for the function and set, this should be good now

**Comment by blegat** [comment]
We are still missing:

- [ ] Implements the getter for ConstraintPrimal/ConstraintPrimalStart and the setter for ConstraintPrimalStart and test them. You can see which bridges already implements it here: https://github.com/JuliaOpt/MathOptInterface.jl/issues/684
- [ ] Implements `MOI.delete` and run `test_delete_bridge`
- [ ] Run `test_model_equals` tests, see https://github.com/JuliaOpt/MathOptInterface.jl/issues/820

**Comment by matbesancon** [comment]
Oh boy `test_delete_bridge` is horrible -_-

**Comment by blegat** [comment]
What's the issue ?

**Comment by matbesancon** [comment]
The signature is rather opaque, errors it yields hard to read

**Comment by matbesancon** [comment]
Ok model equals is working, `test_delete_bridge` is way too hard to decipher for something required to implement bridges

**Comment by blegat** [comment]
Does https://github.com/JuliaOpt/MathOptInterface.jl/pull/954 help ?

**Comment by matbesancon** [comment]
It does yes. Sadly, the test is still failing, the number of variables is 3 at the beginning, 2 after deleting the bridge, all bridged constraints are of number 1, passing at the beginning, failing then

**Comment by blegat** [comment]
The bridge does not seem to implement `ListOfVariableIndices`, `NumberOfVariables`, `ListOfConstraintIndices` and `NumberOfConstraints` which is one thing `test_delete_bridge` is testing.

**Comment by matbesancon** [comment]
Ok good catch, I guess the fallback assumes 0?

**Comment by matbesancon** [comment]
Tests should all be passing now

**Comment by blegat** [comment]
> I guess the fallback assumes 0?

Yes :)

File: test/Bridges/Constraint/indicator_sos.jl [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+    sos1: [w, z] in MathOptInterface.SOS1([0.4, 0.6])
+    wless: w <= 0.0
+    ineq: x + w <= 8.0
+    bin_cons: z in MathOptInterface.ZeroOne()
+    maxobjective: z
+    """
+    model = MOIU.Model{Float64}()
+    MOIU.loadfromstring!(model, s)
+    MOIU.test_models_equal(mock, model, var_names, ["sos1", "wless", "ineq", "bin_cons"])
+
```
Comment (blegat) [inline test/Bridges/Constraint/indicator_sos.jl]: We should also use `test_model_equals` on `bridged_mock`. It also tests that ListOfConstraintIndices, ... are implemented correctly as well as ConstraintFunction, ...

Comment (matbesancon) [reply test/Bridges/Constraint/indicator_sos.jl]: you mean testing against:
```julia
s = """
variables: z, x
indicator_cons: [z, 1.0 * x] in MathOptInterface.IndicatorSet{MathOptInterface.ACTIVATE_ON_ONE}(MathOptInterface.LessThan{Float64}(8.0))
bin_cons: z in MathOptInterface.ZeroOne()
maxobjective: z
"""

```

File: src/Bridges/Constraint/indicator_sos.jl [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    ::Type{<:IndicatorSOS1Bridge})
+    return true
+end
+
+function MOI.get(model::MOI.ModelLike, attr::MOI.ConstraintPrimal, bridge::IndicatorSOS1Bridge)
+    if attr.N == 1
+        return MOI.get(model, MOI.VariablePrimal(), bridge.z_variable_index)
+    end
+    wv = MOI.get(model, MOI.VariablePrimal(), bridge.w_variable_index)
+    lin_primal = MOI.get(model, MOI.ConstraintPrimal(), b.linear_constraint_index)
```
Comment (blegat) [inline src/Bridges/Constraint/indicator_sos.jl]: Replace `MOI.ConstraintPrimal()` by `attr`

File: src/Bridges/Constraint/indicator_sos.jl [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    ::MOI.ConstraintPrimalStart,
+    ::Type{<:IndicatorSOS1Bridge})
+    return true
+end
+
+function MOI.get(model::MOI.ModelLike, attr::MOI.ConstraintPrimal, bridge::IndicatorSOS1Bridge)
+    if attr.N == 1
+        return MOI.get(model, MOI.VariablePrimal(), bridge.z_variable_index)
+    end
+    wv = MOI.get(model, MOI.VariablePrimal(), bridge.w_variable_index)
```
Comment (blegat) [inline src/Bridges/Constraint/indicator_sos.jl]: This should be `MOI.VariablePrimal(attr.N)`

File: src/Bridges/Constraint/indicator_sos.jl [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+function MOI.supports(
+    ::MOI.ModelLike,
+    ::MOI.ConstraintPrimalStart,
+    ::Type{<:IndicatorSOS1Bridge})
+    return true
+end
+
+function MOI.get(model::MOI.ModelLike, attr::MOI.ConstraintPrimal, bridge::IndicatorSOS1Bridge)
+    if attr.N == 1
```
Comment (blegat) [inline src/Bridges/Constraint/indicator_sos.jl]: Why do a special case for `N == 1` ? You should not assume anything on what are the different results and their order

File: src/Bridges/Constraint/indicator_sos.jl (hunk 13 above)
Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: how come? w is a scalar variable

File: src/Bridges/Constraint/indicator_sos.jl (hunk 14 above)
Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: This is getting the `MOI.ConstraintPrimal` for the function [z, a^T x], if this is one, it means getting the primal of a single variable z, otherwise, it's getting the primal of a Scalar Affine

Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: For me this comes from the fact that we are using a vector function, instead of a tuple function: (SingleVariable, AbstractScalarFunction)

File: src/Bridges/Constraint/indicator_sos.jl (hunk 12 above)
Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: done

File: test/Bridges/Constraint/indicator_sos.jl (hunk 11 above)
Comment (matbesancon) [reply test/Bridges/Constraint/indicator_sos.jl]: done

File: src/Bridges/Constraint/indicator_sos.jl (hunk 13 above)
Comment (blegat) [reply src/Bridges/Constraint/indicator_sos.jl]: For the `N`th result, you have `VariablePrimal(N)`, `ConstraintPrimal(N)` and `ConstraintDual(N)`. The value of `ConstraintPrimal(N)` is the value of the function when the variables have values `VariablePrimal(N)`

File: src/Bridges/Constraint/indicator_sos.jl (hunk 14 above)
Comment (blegat) [reply src/Bridges/Constraint/indicator_sos.jl]: This should return a vector of 2 values. `N` is simply the index of the result (i.e. usually it is 1). See `MOI.ResultCount`

File: src/Bridges/Constraint/indicator_sos.jl (hunk 13 above)
Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: ok, that's not really intuitive, I thought `ConstraintPrimal(N)` would be the primal of the `N`-th function in the vector function. I'll update that

File: src/Bridges/Constraint/indicator_sos.jl (hunk 14 above)
Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: so in that case there are the cases `N = 1` and `N = 2`, we can check one and deduce the other, or throw an error but this makes the functions heavier

Comment (blegat) [reply src/Bridges/Constraint/indicator_sos.jl]: Not sure to follow, you just need to make sure that the attributes you create in this function have the same result index. You don't assume anything on the nature of the result index.

Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: oh ok, this is really not obvious eitheir, getting `attr.N` on a function value implies that you are getting the N-th member of a vector-valued function

Comment (blegat) [reply src/Bridges/Constraint/indicator_sos.jl]: Indeed, it is planned to be renamed, see https://github.com/JuliaOpt/MathOptInterface.jl/issues/960

**Comment by matbesancon** [comment]
@blegat after thoughts, not sure `MOI.ConstraintPrimal` should be implemented. You cannot test it without a value set, and you cannot support setting a `ConstraintPrimal` without other things (see error below)

```julia
Getting primal attributes: Error During Test at /home/mbesancon/.julia/dev/MathOptInterface/test/Bridges/Constraint/indicator_sos.jl:179
  Got exception outside of a @test
  ArgumentError: `supports` is not defined for MathOptInterface.ConstraintPrimal(1), it is only defined for attributes such that `is_copyable` returns `true`.
  Stacktrace:
   [1] supports(::MathOptInterface.Utilities.Model{Float64}, ::MathOptInterface.ConstraintPrimal, ::Type{MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}}}) at /home/mbesancon/.julia/dev/MathOptInterface/src/attributes.jl:197
   [2] set(::MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}, ::MathOptInterface.ConstraintPrimal, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.LessThan{Float64}}, ::Float64) at /home/mbesancon/.julia/dev/MathOptInterface/src/Utilities/universalfallback.jl:356

```

**Comment by blegat** [comment]
If you set the variable values, the mock optimizer uses them to compute ConstraintPrimal using `get_fallback`. I dont't why this bridge should not support it.

**Comment by matbesancon** [comment]
Alright `ConstraintPrimal` re-implemented and passing

File: src/Bridges/Constraint/indicator_sos.jl [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+end
+
+function MOI.get(model::MOI.ModelLike, ::MOI.ConstraintPrimal, bridge::IndicatorSOS1Bridge)
+    zvalue = MOI.get(model, MOI.VariablePrimal(), bridge.z_variable_index)
+    wvalue = MOI.get(model, MOI.VariablePrimal(), bridge.w_variable_index)
+    lin_primal_start = MOI.get(model, MOI.ConstraintPrimal(), bridge.linear_constraint_index)
+    return [zvalue, lin_primal_start - wvalue]
+end
+
+function MOI.get(model::MOI.ModelLike, ::MOI.ConstraintPrimalStart, bridge::IndicatorSOS1Bridge)
```
Comment (blegat) [inline src/Bridges/Constraint/indicator_sos.jl]: I would merge this method and the one above into one as it is done in other bridges supporting stating values

Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: In the case of `MOI.ConstraintPrimalStart`, it needs to query `MOI.VariablePrimalStart`, for `MOI.ConstraintPrimal`, it queries MOI.VariablePrimal`. Changing dispatch to reflection with `isa` is not gonna reduce complexity

Comment (blegat) [reply src/Bridges/Constraint/indicator_sos.jl]: Indeed, then you can keep separate methods but add `attr.N` as argument to `VariablePrimal` constructor and use `attr` instead of `MOI.ConstraintPrimal()` otherwise it won't work with multiple results (that are now exposed in JuMP with https://github.com/JuliaOpt/JuMP.jl/pull/2100)

File: src/Bridges/Constraint/indicator_sos.jl [hunk 16]
```diff
@@ -158,26 +158,26 @@ function MOI.supports(
...
+    lin_primal_start = MOI.get(model, attr, bridge.linear_constraint_index)
     return [zvalue, lin_primal_start - wvalue]
 end
 
-function MOI.get(model::MOI.ModelLike, ::MOI.ConstraintPrimalStart, bridge::IndicatorSOS1Bridge)
-    zstart = MOI.get(model, MOI.VariablePrimalStart(), bridge.z_variable_index)
-    wstart = MOI.get(model, MOI.VariablePrimalStart(), bridge.w_variable_index)
-    lin_primal_start = MOI.get(model, MOI.ConstraintPrimalStart(), bridge.linear_constraint_index)
+function MOI.get(model::MOI.ModelLike, attr::MOI.ConstraintPrimalStart, bridge::IndicatorSOS1Bridge)
+    zstart = MOI.get(model, MOI.VariablePrimalStart(attr.N), bridge.z_variable_index)
```
Comment (blegat) [inline src/Bridges/Constraint/indicator_sos.jl]: The start attributes don't have any `result_index` field.

Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: fixed in last commit

File: src/Bridges/Constraint/indicator_sos.jl (hunk 15 above)
Comment (matbesancon) [reply src/Bridges/Constraint/indicator_sos.jl]: Fixed
